### ISSUE TITLE
Convolution weight gradient computation in nn.grad is wrong with groups > 1

### ISSUE THREAD

**Issue by lemonbuffer20** [pytorch/pytorch#51430]
## 🐛 Bug

Convolution weight gradient computation in nn.grad seems to be wrong when groups > 1.
(conv1d_weight, conv2d_weight, conv3d_weight) 

## To Reproduce

Steps to reproduce the behavior: I attached my gist to print the difference between autograd and nn.grad.
In the script, I also suggest a correct alternative (only a minor change was sufficient)

https://gist.github.com/lemonbuffer20/7c98cea1e904b2abaa38ce6d4ec9605e

## Expected behavior

nn.grad.convNd_weight should return the same result as F.conv2d backprop in groups > 1 case.

## Environment

PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce GTX TITAN X

Nvidia driver version: 460.32.03
cuDNN version: Probably one of the following:
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn.so.8
...

## Additional context

1.
I believe the small difference (~=1e-6) of the result is inevitable.
But sometimes the max grad difference (both input and weight) goes to ~=1e-4, which seems much bigger.
Can we remove, or at least reduce the error?

2.
If you can verify my modification is correct, I'd be glad to propose PR.

cc @ezyang @gchanan @zou3519 @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @mruberry @jbschlosser @walterddr @kshitij12345 @bdhirsh

**Comment by gchanan** [comment]
is this the same as https://github.com/pytorch/pytorch/issues/50042?

**Comment by lemonbuffer20** [comment]
@gchanan No, I think that's a different issue. The mentioned issue is related to nn.Conv2d gradient calculate behavior (which depends on CPU backward kernel), and this issue is about manual computation of gradient in nn.grad.

**Comment by zou3519** [comment]
These are for the "slow" backward formulas we have at https://github.com/pytorch/pytorch/blob/b71215a9097cbd0170e07db6f89ea9fdd932c4a5/torch/nn/grad.py#L170. @lemonbuffer20 what is your use case for using those formulas as opposed to using autograd to compute gradients?

**Comment by lemonbuffer20** [comment]
@zou3519 I was trying to modify convolution backward, specificially, changing batch-wise aggregation (usually sum). Common nn.Conv2d or cudnn_backward does not support such figure, so I tried to implement custom function with nn.grad.conv2d_backward,

**Comment by geohot** [comment]
Can confirm this is true, and wasted 2 hours on it.

The initial repeat is wrong, as it's also repeating the groups. Line 35 in that commit is the fix.

See my fix in my neural network library here: https://github.com/geohot/tinygrad/commit/8864b373338886a9173d3f823154815535104f28

**Comment by Lezcano** [comment]
cc @nikitaved could you have a look at this one?

**Comment by jbschlosser** [comment]
I'll look into rewriting the back of these APIs to use the general `convolution_backward` op now that it exists. That should remove any discrepancies and fix this bug.
